packages/gatsby/src/state-machines/develop/index.ts

```diff
@@ -145,28 +153,34 @@ const developConfig: MachineConfig<IBuildContext, any, AnyEventObject> = {
             target: `recreatingPages`,
             cond: ({ nodesMutatedDuringQueryRun }: IBuildContext): boolean =>
               !!nodesMutatedDuringQueryRun,
-            actions: [`markNodesClean`, `incrementRecompileCount`],
+            actions: [
+              `markNodesClean`,
+              `incrementRecompileCount`,
+              `clearPendingQueryRuns`,
+            ],
           },
           {
             // If we have no compiler (i.e. it's first run), then spin up the
             // webpack and socket.io servers
             target: `startingDevServers`,
-            actions: `setQueryRunningFinished`,
+            actions: [`setQueryRunningFinished`, `clearPendingQueryRuns`],
             cond: ({ compiler }: IBuildContext): boolean => !compiler,
           },
           {
             // If source files have changed, then recompile the JS bundle
             target: `recompiling`,
             cond: ({ sourceFilesDirty }: IBuildContext): boolean =>
               !!sourceFilesDirty,
+            actions: [`clearPendingQueryRuns`],
           },
           {
             // ...otherwise just wait.
             target: `waiting`,
+            actions: [`clearPendingQueryRuns`],
           },
         ],
         onError: {
-          actions: `logError`,
+          actions: [`logError`, `clearPendingQueryRuns`],
```

Comment:
not sure if we should clear those in `onError` - probably not? We don't know where we failed so there is a chance there are `page-data` requests waiting for results that will never come? Should I track pending requests and possibly fail those on errors?

Reply:
Possibly yeah, I have no idea what can cause for this to result in `onError`, so if this fails in a way that fail every time and we don't clear we would end up in infinite loop - `query-running` -> error -> `waiting` -> `query-running` (because we didn't clear).

So failing express requests (`500` error?) might be more reasonable default for the whole umbrella of errors that can happen inside `query-running` state?

Reply:
I agree clearing it is the best idea

Reply:
Would it be acceptable to keep this as-is for now and add additional handling here? I think if something fails in this experimental feature we should get more ideas when we get reports about it to know how to handle it?

The whole `pendingQueryRuns` is only used when Query on Demand is enabled - we won't emit it in regular `gatsby develop` mode so it won't impact users who don't opt-in into experimental feature

Reply:
Agree. Let's keep it as is